src/wrappers/OsipiBase.py

```diff
@@ -83,11 +83,77 @@ def osipi_fit(self, data, bvalues, **kwargs):
         
         #args = [data, use_bvalues, use_initial_guess, use_bounds, use_thresholds]
         #args = [arg for arg in args if arg is not None]
-        results = self.ivim_fit(*args, **kwargs)
+
+        # Check if there is an attribute that defines the result dictionary keys
+        if hasattr(self, "result_keys"):
+            # result_keys is a list of strings of parameter names, e.g. "S0", "f1", "f2", etc.
+            result_keys = self.result_keys
+        else:
+            # Default is ["f", "D*", "D"]
+            self.result_keys = ["f", "D*", "D"]
+        
+        results = {}
+        for key in self.result_keys:
+            results[key] = np.empty(list(data.shape[:-1]))
+
+        # Assuming the last dimension of the data is the signal values of each b-value
+        #results = np.empty(list(data.shape[:-1])+[3]) # Create an array with the voxel dimensions + the ones required for the fit
+        #for ijk in tqdm(np.ndindex(data.shape[:-1]), total=np.prod(data.shape[:-1])):
+            #args = [data[ijk], use_bvalues]
+            #fit = list(self.ivim_fit(*args, **kwargs))
+            #results[ijk] = fit
+
+        for ijk in tqdm(np.ndindex(data.shape[:-1]), total=np.prod(data.shape[:-1])):
```

Comment:
I've been thinking about parallelizing the fitting and have something similar running as a separate script. I'm thinking that this structure is fine for now but will be too rigid. I think we'd want each algorithm to determine what is input into each step. Perhaps some combine all directions in one input, for example. Perhaps there's a generic generator that does essentially this, but can be overridden by the algorithm to supply data as it sees fit.

I should just finish up my changes and push them maybe it would make more sense to have it all in context.

Reply:
I was working on fixing the notebook yesterday evening and ran into the volume being too large and fitting would take 1 hour if I didn't use Paulien and Merels selective indexing (I tried to avoid the 1D data reformatting). I agree that a loop like this would be a good place for some parallelization, although I would like to not be the one doing it as it has been a great source of headaches in the past :)

But regarding multi-directional data, the only way I've seen and handled such data has been as a long list of b-values with separate header-info to discriminate the directions. And it is sort of the easiest way to do it. I think DIPY essentially forces you to do it that way to avoid these formatting issues.

I'm gonna push my fixes and we could hold off on this merge until you've pushed yours?

Reply:
Hey! My commits already do parallel fits. I'm not sure whether it is the most efficient implementation, but it is a simple one. 

https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/blob/44051e4f978e26df885891180d7350bc86733e02/src/original/OGC_AmsterdamUMC/LSQ_fitting.py#L87-L110

    # here we try parallel computing, but if fails, go back to computing one single core.
    single = False
    if njobs > 2:
        try:
            # define the parallel function
            def parfun(i):
                return fit_segmented(bvalues, dw_data[i, :], bounds=bounds, cutoff=cutoff,p0=p0)


            output = Parallel(n_jobs=njobs)(delayed(parfun)(i) for i in tqdm(range(len(dw_data)), position=0, leave=True))
            Dt, Fp, Dp = np.transpose(output)
        except:
            # if fails, retry using single core
            single = True
    else:
        # or, if specified, immediately go to single core
        single = True
    if single:
        # initialize empty arrays
        Dp = np.zeros(len(dw_data))
        Dt = np.zeros(len(dw_data))
        Fp = np.zeros(len(dw_data))
        for i in tqdm(range(len(dw_data)), position=0, leave=True):
            # fill arrays with fit results on a per voxel base:
            Dt[i], Fp[i], Dp[i] = fit_segmented(bvalues, dw_data[i, :], bounds=bounds, cutoff=cutoff,p0=p0)